File: Day47/Part1/mirror_boss.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>
#include <set>

using namespace std;

struct State {
    int px, py, bx, by;
    
    bool operator<(const State& other) const {
        return tie(px, py, bx, by) < tie(other.px, other.py, other.bx, other.by);
    }
};

int main() {
    vector<vector<int>> grid = {
        {0, 0, 0, 0, 0, 0},
        {0, 0, 0, 0, 0, 0},
        {0, 0, 0, 0, 0, 0},
        {0, 0, 0, 0, 0, 0},
        {0, 0, 0, 0, 0, 0},
        {0, 0, 0, 0, 0, 0}
    };
    int N = grid.size();
    int M = grid[0].size();
    
    int sx = 0, sy = 0;
    int bx = 5, by = 5;
    int tx = 5, ty = 5;
    
    queue<pair<int, State>> q;
    set<State> visited;
    
    State startState = {sx, sy, bx, by};
    q.push({0, startState});
    visited.insert(startState);
    
    int dx[] = {-1, 1, 0, 0};
    int dy[] = {0, 0, -1, 1};
    
    while (!q.empty()) {
        auto [steps, s] = q.front();
        q.pop();
        
        // cout << "Step " << steps << ": P(" << s.px << "," << s.py << ") B(" << s.bx << "," << s.by << ")" << endl;

        if (s.px == tx && s.py == ty) {
            cout << "Minimum Steps: " << steps << endl;
            return 0;
        }
        
        for (int i = 0; i < 4; ++i) {
            int npx = s.px + dx[i];
            int npy = s.py + dy[i];
            
            int nbx = s.bx - dx[i];
            int nby = s.by - dy[i];
            
            // Check Player
            if (npx >= 0 && npx < N && npy >= 0 && npy < M && grid[npx][npy] == 0) {
                // Check Boss
                if (nbx >= 0 && nbx < N && nby >= 0 && nby < M && grid[nbx][nby] == 0) {
                    State nextState = {npx, npy, nbx, nby};
                    if (visited.find(nextState) == visited.end()) {
                        visited.insert(nextState);
                        q.push({steps + 1, nextState});
                    }
                }
            }
        }
    }
    
    cout << "-1" << endl;
    return 0;
}

File: learnLeetCode/Day47/Part1/mirror_boss.py
import collections

def solve():
    # Hardcoded Input
    # 6x6 Grid
    grid = [
        [0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0]
    ]
    N = len(grid)
    M = len(grid[0])
    
    start_player = (0, 0)
    start_boss = (5, 5)
    target = (5, 5)
    
    # BFS
    # State: (px, py, bx, by)
    initial_state = (start_player[0], start_player[1], start_boss[0], start_boss[1])
    queue = collections.deque([(0, initial_state)])
    visited = set([initial_state])
    
    moves = [(-1, 0), (1, 0), (0, -1), (0, 1)] # Up, Down, Left, Right
    
    while queue:
        steps, (px, py, bx, by) = queue.popleft()
        
        if (px, py) == target:
            print(f"Minimum Steps: {steps}")
            return
        
        for dx, dy in moves:
            # Player move
            npx, npy = px + dx, py + dy
            # Boss move (Mirrored)
            nbx, nby = bx - dx, by - dy
            
            # Check bounds and walls for Player
            if 0 <= npx < N and 0 <= npy < M and grid[npx][npy] == 0:
                # Check bounds and walls for Boss
                if 0 <= nbx < N and 0 <= nby < M and grid[nbx][nby] == 0:
                    new_state = (npx, npy, nbx, nby)
                    if new_state not in visited:
                        visited.add(new_state)
                        queue.append((steps + 1, new_state))
                        
    print("-1")

if __name__ == "__main__":
    solve()
